File: backend/app/platform/feature_registry.py
"""Central feature registry for analytics modules and platform surfaces."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from app.auth.models import Permission


FilterKey = Literal["date", "region", "category", "customer", "product", "segment", "department", "supplier", "search"]


@dataclass(frozen=True)
class FeatureDefinition:
    id: str
    route: str
    api_prefix: str | None
    navigation_label: str
    icon: str
    permissions: tuple[Permission, ...]
    supported_filters: tuple[FilterKey, ...] = ()
    export_support: bool = False
    available: bool = True
    section: str = "main"
    description: str = ""
    keywords: tuple[str, ...] = ()
    tags: tuple[str, ...] = ()


FEATURE_REGISTRY: tuple[FeatureDefinition, ...] = (
    FeatureDefinition(
        id="dashboard",
        route="/dashboard",
        api_prefix="/api/v1/dashboard",
        navigation_label="Dashboard",
        icon="LayoutDashboard",
        permissions=(Permission.DASHBOARD_READ,),
        supported_filters=("date",),
        export_support=False,
        available=True,
        description="Executive performance overview",
        keywords=("home", "overview"),
        tags=("executive",),
    ),
    FeatureDefinition(
        id="sales",
        route="/sales",
        api_prefix="/api/v1/sales",
        navigation_label="Sales Intelligence",
        icon="ShoppingCart",
        permissions=(Permission.SALES_READ,),
        supported_filters=("date", "region", "category", "search"),
        export_support=True,
        available=True,
        description="Commercial revenue and product performance",
        keywords=("revenue", "orders"),
        tags=("sales",),
    ),
    FeatureDefinition(
        id="customers",
        route="/customers",
        api_prefix="/api/v1/customers",
        navigation_label="Customer Intelligence",
        icon="Users",
        permissions=(Permission.CUSTOMERS_READ,),
        supported_filters=("date", "region", "segment", "search"),
        export_support=True,
        available=True,
        description="Lifecycle, RFM, cohorts, and churn",
        keywords=("segment", "churn"),
        tags=("customers",),
    ),
    FeatureDefinition(
        id="finance",
        route="/finance",
        api_prefix="/api/v1/finance",
        navigation_label="Finance Intelligence",
        icon="Wallet",
        permissions=(Permission.FINANCE_READ,),
        supported_filters=("date", "region", "department", "category", "search"),
        export_support=True,
        available=True,
        description="Profitability, costs, cashflow, and budget",
        keywords=("margin", "profit"),
        tags=("finance",),
    ),
    FeatureDefinition(
        id="operations",
        route="/operations",
        api_prefix="/api/v1/operations",
        navigation_label="Operations Intelligence",
        icon="Package",
        permissions=(Permission.OPERATIONS_READ,),
        supported_filters=("date", "region", "category", "supplier", "search"),
        export_support=True,
        available=True,
        description="Inventory, suppliers, returns, and warehouses",
        keywords=("inventory", "fulfillment"),
        tags=("operations",),
    ),
    FeatureDefinition(
        id="analytics",
        route="/analytics",
        api_prefix=None,
        navigation_label="Analytics",
        icon="BarChart3",
        permissions=(Permission.ANALYTICS_READ,),
        supported_filters=("date", "search"),
        export_support=False,
        available=False,
        description="Cross-module analytics explorer (coming soon)",
        keywords=("explore", "insights"),
        tags=("analytics",),
    ),
    FeatureDefinition(
        id="predictions",
        route="/predictions",
        api_prefix=None,
        navigation_label="AI Predictions",
        icon="Brain",
        permissions=(Permission.ML_READ,),
        supported_filters=("date", "search"),
        export_support=False,
        available=False,
        description="Machine learning predictions (coming soon)",
        keywords=("ml", "forecast", "models"),
        tags=("ml",),
    ),
    FeatureDefinition(
        id="recommendations",
        route="/recommendations",
        api_prefix=None,
        navigation_label="Business Recommendations",
        icon="Sparkles",
        permissions=(Permission.ANALYTICS_READ,),
        export_support=False,
        available=False,
        description="Decision recommendations (coming soon)",
        keywords=("decisions", "actions"),
        tags=("recommendations",),
    ),
    FeatureDefinition(
        id="reports",
        route="/reports",
        api_prefix=None,
        navigation_label="Reports",
        icon="FileText",
        permissions=(Permission.REPORTS_READ,),
        export_support=True,
        available=False,
        description="Report center (coming soon)",
        keywords=("export", "pdf"),
        tags=("reports",),
    ),
    FeatureDefinition(
        id="settings",
        route="/settings",
        api_prefix=None,
        navigation_label="Settings",
        icon="Settings",
        permissions=(Permission.SETTINGS_READ,),
        available=True,
        section="system",
        description="Workspace preferences",
        keywords=("preferences", "profile"),
        tags=("system",),
    ),
)


def list_features(*, available_only: bool = False) -> list[FeatureDefinition]:
    features = list(FEATURE_REGISTRY)
    if available_only:
        features = [f for f in features if f.available]
    return features


def get_feature(feature_id: str) -> FeatureDefinition | None:
    for feature in FEATURE_REGISTRY:
        if feature.id == feature_id:
            return feature
    return None


def feature_for_api_path(path: str) -> FeatureDefinition | None:
    matches = [
        f
        for f in FEATURE_REGISTRY
        if f.api_prefix and (path == f.api_prefix or path.startswith(f"{f.api_prefix}/"))
    ]
    if not matches:
        return None
    return sorted(matches, key=lambda f: len(f.api_prefix or ""), reverse=True)[0]


def required_permissions_for_path(path: str) -> tuple[Permission, ...]:
    feature = feature_for_api_path(path)
    if feature is None:
        return ()
    return feature.permissions

File: frontend/src/config/feature-registry.ts
/**
 * Central frontend feature registry — mirrors backend platform registry.
 */

import type { LucideIcon } from "lucide-react";
import {
  BarChart3,
  Brain,
  FileText,
  LayoutDashboard,
  Package,
  Settings,
  ShoppingCart,
  Sparkles,
  Users,
  Wallet,
} from "lucide-react";

export type FeaturePermission =
  | "dashboard:read"
  | "sales:read"
  | "customers:read"
  | "finance:read"
  | "operations:read"
  | "analytics:read"
  | "ml:read"
  | "reports:read"
  | "settings:read"
  | "settings:write"
  | "export:run"
  | "admin:all";

export type FeatureFilterKey =
  | "date"
  | "region"
  | "category"
  | "customer"
  | "product"
  | "segment"
  | "department"
  | "supplier"
  | "search";

export type FeatureDefinition = {
  id: string;
  route: string;
  apiPrefix?: string;
  navigationLabel: string;
  icon: LucideIcon;
  permissions: FeaturePermission[];
  supportedFilters: FeatureFilterKey[];
  exportSupport: boolean;
  available: boolean;
  section: "main" | "system";
  description?: string;
  keywords?: string[];
};

export const featureRegistry: FeatureDefinition[] = [
  {
    id: "dashboard",
    route: "/dashboard",
    apiPrefix: "/api/v1/dashboard",
    navigationLabel: "Dashboard",
    icon: LayoutDashboard,
    permissions: ["dashboard:read"],
    supportedFilters: ["date"],
    exportSupport: false,
    available: true,
    section: "main",
    keywords: ["home", "overview"],
  },
  {
    id: "sales",
    route: "/sales",
    apiPrefix: "/api/v1/sales",
    navigationLabel: "Sales Intelligence",
    icon: ShoppingCart,
    permissions: ["sales:read"],
    supportedFilters: ["date", "region", "category", "search"],
    exportSupport: true,
    available: true,
    section: "main",
    keywords: ["revenue", "orders"],
  },
  {
    id: "customers",
    route: "/customers",
    apiPrefix: "/api/v1/customers",
    navigationLabel: "Customer Intelligence",
    icon: Users,
    permissions: ["customers:read"],
    supportedFilters: ["date", "region", "segment", "search"],
    exportSupport: true,
    available: true,
    section: "main",
    keywords: ["segment", "churn"],
  },
  {
    id: "finance",
    route: "/finance",
    apiPrefix: "/api/v1/finance",
    navigationLabel: "Finance Intelligence",
    icon: Wallet,
    permissions: ["finance:read"],
    supportedFilters: ["date", "region", "department", "category", "search"],
    exportSupport: true,
    available: true,
    section: "main",
    keywords: ["margin", "profit"],
  },
  {
    id: "operations",
    route: "/operations",
    apiPrefix: "/api/v1/operations",
    navigationLabel: "Operations Intelligence",
    icon: Package,
    permissions: ["operations:read"],
    supportedFilters: ["date", "region", "category", "supplier", "search"],
    exportSupport: true,
    available: true,
    section: "main",
    keywords: ["inventory", "fulfillment"],
  },
  {
    id: "analytics",
    route: "/analytics",
    navigationLabel: "Analytics",
    icon: BarChart3,
    permissions: ["analytics:read"],
    supportedFilters: ["date", "search"],
    exportSupport: false,
    available: false,
    section: "main",
    keywords: ["explore", "insights"],
  },
  {
    id: "predictions",
    route: "/predictions",
    navigationLabel: "AI Predictions",
    icon: Brain,
    permissions: ["ml:read"],
    supportedFilters: ["date", "search"],
    exportSupport: false,
    available: false,
    section: "main",
    keywords: ["ml", "forecast", "models"],
  },
  {
    id: "recommendations",
    route: "/recommendations",
    navigationLabel: "Business Recommendations",
    icon: Sparkles,
    permissions: ["analytics:read"],
    supportedFilters: [],
    exportSupport: false,
    available: false,
    section: "main",
    keywords: ["decisions", "actions"],
  },
  {
    id: "reports",
    route: "/reports",
    navigationLabel: "Reports",
    icon: FileText,
    permissions: ["reports:read"],
    supportedFilters: [],
    exportSupport: true,
    available: false,
    section: "main",
    keywords: ["export", "pdf"],
  },
  {
    id: "settings",
    route: "/settings",
    navigationLabel: "Settings",
    icon: Settings,
    permissions: ["settings:read"],
    supportedFilters: [],
    exportSupport: false,
    available: true,
    section: "system",
    keywords: ["preferences", "profile"],
  },
];

export function listAvailableFeatures(): FeatureDefinition[] {
  return featureRegistry.filter((f) => f.available);
}

export function getFeature(id: string): FeatureDefinition | undefined {
  return featureRegistry.find((f) => f.id === id);
}

export function canAccessFeature(
  feature: FeatureDefinition,
  permissions: readonly string[],
): boolean {
  if (!feature.available) return false;
  if (permissions.includes("admin:all")) return true;
  if (feature.permissions.length === 0) return true;
  return feature.permissions.some((p) => permissions.includes(p));
}

export function filterFeaturesByPermissions(
  permissions: readonly string[],
  options: { includeUnavailable?: boolean } = {},
): FeatureDefinition[] {
  const includeUnavailable = options.includeUnavailable ?? false;
  return featureRegistry.filter((feature) => {
    if (!includeUnavailable && !feature.available) return false;
    if (!feature.available && includeUnavailable) {
      return permissions.includes("admin:all")
        || feature.permissions.some((p) => permissions.includes(p));
    }
    return canAccessFeature(feature, permissions);
  });
}
